### PR TITLE
Environment variables support

### PR DESCRIPTION
**Pull request by eps90**
Hi devs!
Recently I've noticed there is abandoned #160 issue, however I think this is gonna be useful for many of Magallanes users. 

* Environent variables can be set by passing options to general config...
```yaml
# .mage/config/general.yml
name: My fantastic App
email: andresmontanez@gmail.com
env:
  variables:  
    node_path: "/bin/node"
    symfony_env: prod
```

* ... or environment config...
```yaml
# .mage/config/environment/production.yml
deployment:
  user: root
  from: ./
  to: /var/www/vhosts/example.com/www
env:
  variables:
    symfony_env: prod
```
* ... or even task config!
```yaml
# .mage/config/environment/production.yml
tasks:
  pre-deploy:
    - scm/update
  on-deploy:
    - symfony2/cache-warmup: { env: prod, env: { variables: {env1: key1} } }
  # short way
    - symfony2/cache-warmup: { env: prod, env.variables: {env1: key1} }
```

When you run the command, the set variables are prepended to the original command:
```
$ SYMFONY_ENV=prod composer install
```
(**Note**: The script does not uppercases variables, they're gonna be passed in the same way you typed them in config)

Given variables have the following priorities:

* Task parameters and command line parameters
* Environment config parameters
* General config parameters

That means, if you have the env variable in your task configuration and there's an env var in your environment configuration, the task's one will override the environment's config.

The scripts allows also to prepend all system's environment variables, but it needs proper setting in your `php.ini` for CLI:
```ini
# php.ini
variables_order="EGPCS"
```

Then, you can set your settings from `.bashrc` or `.zshrc`, like:
```
# .zshrc
export NODE_PATH="bin/node"
```

System's env variables has the lowest prority to give you ability to override them in config.

**Warning:** Including system's environment variables can slow down the applicattion, so use it carefully.
Feel free to review and comment! I know there's gonna be do much more, all sugestions are welcome!

### CODE REVIEW

File: Mage/Task/AbstractTask.php [hunk 1]
```diff
@@ -307,4 +308,53 @@ protected function untarRelease($releaseId)
...
+    protected function getEnvVariables()
+    {
+        $configVars = array_merge(
+            $this->getConfig()->general('env', []),
+            $this->getConfig()->environmentConfig('env', []),
+            $this->getConfig()->getParameter('env', [])
+        );
+
+        if (isset($configVars['variables'])) {
+            $configVars = $configVars['variables'];
```
Comment (jhuet) [inline Mage/Task/AbstractTask.php]: If we do that, will variables like ``- symfony2/cache-warmup: { env: prod, env: {env1: key1} }`` still be available if we also have ``env: variables: ...`` in ``general.yml`` ?

Comment (eps90) [inline Mage/Task/AbstractTask.php]: Um yeah, you're right. I've mistyped this in docs.

Comment (eps90) [inline Mage/Task/AbstractTask.php]: Ok, just fixed, what do think about it?

Comment (eps90) [inline Mage/Task/AbstractTask.php]: But yes, as I wrote in description, all configs are merged. So you can have your vars in both places.

**Comment by eps90** [comment]
@nicolas-san @Landkeks Guys, can you take a look at that PR? Looks like you've posted a feature request for this issue :)

**Comment by nicolas-san** [comment]
You have been faster than me ^^
I'v to setup new work project, i'll try your code locally next wednesday

**Comment by Landkeks** [comment]
Looks good, although I haven't tested it yet. I'll try it as soon as I need to change anything on my site. Thanks a lot!

**Comment by nicolas-san** [comment]
Almost ready to test, I think tonight or tomorrow, I test with a config env in environment file

**Comment by eps90** [comment]
Great! :)

**Comment by nicolas-san** [comment]
Ok
tested this day, no problem, with conf in env config.
I can mail .phar archive to anybody want to test it

**Comment by eps90** [comment]
Awesome :) So let's @Landkeks repond us what he thinks about it :)

**Comment by Landkeks** [comment]
I'm currently busy with a lot of other stuff, so I'm not going to be able to test this anytime soon, sorry! But, as I said above, the PR looks good to me.

**Comment by eps90** [comment]
@Landkeks Sure, don't worry.

**Comment by MadMind** [comment]
Any updates on this?

**Comment by eps90** [comment]
@MadMind Actually I think it can be merged and then wait for feedback.

**Comment by nicolas-san** [comment]
@eps90 @MadMind I use it since my comment, with no issue

**Comment by MadMind** [comment]
Unfortunately it does not work in my case :(
I want to pass GIT_SSH to remote host. I've added some comments int the merged commit..

**Comment by eps90** [comment]
Thanks @MadMind , I'll take a look on that in the evening. Your comment could be helpful here, because I didn't check it with `git-rebase` strategy.

**Comment by drAlberT** [comment]
:+1:   I'm particularly interested in 

```
env:
  variables:
    symfony_env: prod 
```

so we can avoid to repeat `{env: prod }` everywhere

File: Mage/Task/AbstractTask.php [hunk 2]
```diff
@@ -307,4 +308,56 @@ protected function untarRelease($releaseId)
...
+     * @see http://php.net/manual/en/ini.core.php#ini.variables-order
+     *
+     * @return array
+     */
+    protected function getEnvVariables()
+    {
+        $configVars = array_merge(
+            $this->getConfig()->general('env', []),
+            $this->getConfig()->environmentConfig('env', []),
+            $this->getConfig()->getParameter('env', []),
```
Comment (jhuet) [inline Mage/Task/AbstractTask.php]: This line should be : `$this->getParameter('env', array())` in order to retrieve actual task config parameter. Same goes for line 327 for the `env.variables` shortcut.

Comment (drAlberT) [inline Mage/Task/AbstractTask.php]: right

**Pull request merged**